### PR TITLE
time: fix the fraction loss when encountering time string with timezone suffix

### LINKED ISSUE #35291: time: insert item lost the fraction when encountering time string with timezone suffix
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
DROP TABLE IF EXISTS `testSetObjectTime`;
CREATE TABLE `testSetObjectTime` (
  `id` int(11) DEFAULT NULL,
  `d` time(6) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
INSERT INTO testSetObjectTime VALUES (1, '2020-01-01 12:00:00.123456000+05:00');
INSERT INTO testSetObjectTime VALUES (2, '2020-01-01 12:00:00.123456+05:00');
SELECT * FROM testSetObjectTime;
```

### 2. What did you expect to see? (Required)
MySQL has its warnings as `Incorrect time value: '2020-01-01 12:00:00.123456+05:00' for column 'd' at row 1`
So it won't recognize the timezone suffix and convert the `12:00:00.123456` to its default `UTC` timezone as `07:00:00.123456`

But TiDB will convert the `12:00:00.123456` (from source +5:00, yes we recognized it and used it as source time zone) to its default `Asia/Shanghai timezone` (+8:00) as `15:00:00.123456`, but unfortunately, we lost the fraction here.
```
+------+-----------------+
| id   | d               |
+------+-----------------+
|    1 | 15:00:00.123456 |
|    2 | 15:00:00.123456 |
+------+-----------------+
```

### 3. What did you see instead (Required)
```
+------+-----------------+
| id   | d               |
+------+-----------------+
|    1 | 15:00:00.000000 |
|    2 | 15:00:00.000000 |
+------+-----------------+
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by Qiaolin-Yu**
### What problem does this PR solve?

Issue Number: close #35291

Problem Summary:

### What is changed and how it works?
Identify the position of the timezone before identifying the position of the fraction in the time string to avoid the accuracy loss. Timezone will also be checked later to ensure compatibility so that there is duplicate detection for timezone.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [x] No code

```
DROP TABLE IF EXISTS `testSetObjectTime`;
CREATE TABLE `testSetObjectTime` (
  `id` int(11) DEFAULT NULL,
  `d` time(6) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
INSERT INTO testSetObjectTime VALUES (1, '2020-01-01 12:00:00.123456000+05:00');
INSERT INTO testSetObjectTime VALUES (2, '2020-01-01 12:00:00.123456+05:00');
SELECT * FROM testSetObjectTime;
```

```
+------+-----------------+
| id   | d               |
+------+-----------------+
|    1 | 15:00:00.123456 |
|    2 | 15:00:00.123456 |
+------+-----------------+
```
Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -576,8 +576,15 @@ func GetFsp(s string) int {
 // GetFracIndex finds the last '.' for get fracStr, index = -1 means fracStr not found.
 // but for format like '2019.01.01 00:00:00', the index should be -1.
 func GetFracIndex(s string) (index int) {
+	tzIndex, _, _, _, _ := GetTimezone(s)
+	var end int
+	if tzIndex != -1 {
+		end = tzIndex
```
Comment (bestwoody) [inline types/time.go:582]: is char `s[tzIndex]` a part of datetime? since the for following for-loop begin at `tzIndex` not `tzIndex-1`

Comment (bestwoody) [reply types/time.go]: have you test it locally？

File: types/time_test.go [hunk 2]
```diff
@@ -110,6 +110,9 @@ func TestDateTime(t *testing.T) {
 		{"2020-05-28 23:59:59T T00:00:00", "2020-05-28 23:59:59"},
 		{"2020-10-22 10:31-10:12", "2020-10-22 10:31:10"},
 		{"2018.01.01 01:00:00", "2018-01-01 01:00:00"},
+
+		// For issue 35291
+		{"2020-01-01 12:00:00.123456+05:00", "2020-01-01 07:00:00.123456"},
```
Comment (bestwoody) [inline types/time_test.go:115]: why the test can be passed when the line `end=index` not `index-1`  is before？

**Review by XuHuaiyu — APPROVED** [review]

File: types/time.go (hunk 1 above)
Comment (Qiaolin-Yu) [reply types/time.go]: yes

File: types/time_test.go (hunk 2 above)
Comment (Qiaolin-Yu) [reply types/time_test.go:115]: Because  ```unicode.IsPunct(':')``` is ```true```, and ```unicode.IsPunct('+')``` is ```false```.

**Review by bestwoody — APPROVED** [review]
lgtm

**Review by LittleFall — APPROVED** [review]

**Pull request merged**